### PR TITLE
Bluetooth: Mesh: Missed IV update cannot be captured

### PR DESCRIPTION
**Pull request by Balaklaka**
It seems that if the IV update is missed, a node cannot
recover it until the IV index has increased to a value
greater than Node's Last known IV + 1.

Signed-off-by: Ingar Kulbrandstad <ingar.kulbrandstad@nordicsemi.no>

### CODE REVIEW

File: subsys/bluetooth/mesh/net.c [hunk 1]
```diff
@@ -266,14 +266,26 @@ bool bt_mesh_net_iv_update(uint32_t iv_index, bool iv_update)
...
+			 * node is in Normal mode of operation, it means
+			 * previous IV update has been missed, immediately
+			 * update the IV Index without going to IV update in
+			 * progress state.
+			 */
+			BT_WARN("Performing IV Index Recovery");
+			bt_mesh_rpl_clear();
+			bt_mesh.iv_index = iv_index;
+			bt_mesh.seq = 0U;
+			goto do_update;
```
Comment (trond-snekvik) [inline subsys/bluetooth/mesh/net.c:288]: Can this be combined with lines 260-266, perhaps? e.g. change line 260 to `if (iv_index > bt_mesh.iv_index)`, then insert the guard on lines 268-276 on line 261

**Review by trond-snekvik — COMMENTED** [review]
Relevant spec text (section 3.10.5):
> If a node in Normal Operation receives a Secure Network beacon with an IV index equal to the last known IV index+1 and the IV Update Flag set to 0, the node may update its IV without going to the IV Update in Progress state, or it may initiate an IV Index Recovery procedure (Section 3.10.6), or it may ignore the Secure Network beacon. The node makes the choice depending on the time since last IV update and the likelihood that the node has missed the Secure Network beacons with the IV update Flag set to 1.

File: subsys/bluetooth/mesh/net.c [hunk 2]
```diff
@@ -266,14 +266,26 @@ bool bt_mesh_net_iv_update(uint32_t iv_index, bool iv_update)
 		}
 
 		if (iv_index == bt_mesh.iv_index + 1 && !iv_update) {
-			BT_WARN("Ignoring new index in normal mode");
-			return false;
-		}
+			if (!(IS_ENABLED(CONFIG_BT_MESH_IV_UPDATE_TEST) &&
+			      atomic_test_bit(bt_mesh.flags, BT_MESH_IVU_TEST))) {
+				if (bt_mesh.ivu_duration < BT_MESH_IVU_MIN_HOURS) {
```
[the remark this answers is not in the document]
Comment (omkar3141) [reply subsys/bluetooth/mesh/net.c:271]: Also, I'm wondering why are these lines added? I think only Line 268 to Line 271, were needed to be removed to enable IV index migration upon missed IV update.

[the remark this answers is not in the document]
Comment (alxelax) [reply subsys/bluetooth/mesh/net.c:271]: I guess it is because the requirement to follow 96 hours rule is mandatory.

[the remark this answers is not in the document]
Comment (omkar3141) [reply subsys/bluetooth/mesh/net.c:271]: The requirement does not apply in the case as stated in the paragraph in the Jira ticket.

[the remark this answers is not in the document]
Comment (Balaklaka) [reply subsys/bluetooth/mesh/net.c:271]: Do we need to change this at all. The spec allow to choose  any of the 3 solutions, as it uses words as "may" and "or". The existing solution chose the last solution. This has been this way for several years. If we should change it it was not enough to do it as it said in the issue, because that would have given the same result with exit in the "!iv_upate" bellow.
The time (96 hours) was chosen as that is the only thing we have to go on. I will update with a  local variable.

[the remark this answers is not in the document]
Comment (omkar3141) [reply subsys/bluetooth/mesh/net.c:271]: Yes, we still need a change, because if we don't do this there is no way for the device to recover by itself (and we don't provide any way that takes care of other options provided by "may") resulting in the device getting thrown out of the network and becoming unrechable.

[the remark this answers is not in the document]
Comment (Balaklaka) [reply subsys/bluetooth/mesh/net.c]: After discussion with @omkar3141, I disregarded the part about "The node makes the choice depending on the time since IV update and the likelihood that the node has missed the Secure Network beacons with the IV update Flag set to 1." as this does not specify any time. Also found that it is better to do a "IV Index Recovery procedure" than the "ignore the Secure Network beacon". This way the node will continue to function after a missed IV update. This also make this function the same way as if IV index is greater than IV index + 1.

[the remark this answers is not in the document]
Comment (alxelax) [reply subsys/bluetooth/mesh/net.c]:
```
3.10.5 IV Update procedure
When a node is added to a network, the node is given an IV Index. If the node is added to a network when the network is in Normal operation, then it shall operate in Normal operation for at least 96 hours. If a node is added to a network while the network is in the IV Update in Progress state, then the node shall be given the new IV Index value and operate in Normal operation for at least 96 hours.
```

I don't see any reason why this rule should be ignored here just basing on the assumption the device was thrown out of the network. It doesn't seem correct.

File: subsys/bluetooth/mesh/net.c [hunk 3]
```diff
@@ -257,24 +257,19 @@ bool bt_mesh_net_iv_update(uint32_t iv_index, bool iv_update)
 			return false;
 		}
 
-		if (iv_index > bt_mesh.iv_index + 1) {
+		if ((iv_index > bt_mesh.iv_index + 1) ||
+		    (iv_index == bt_mesh.iv_index + 1 && !iv_update)) {
```
Comment (alxelax) [inline subsys/bluetooth/mesh/net.c:261]: this condition requires to be 'and' with 96 hours rule checking

**Review by alxelax — CHANGES_REQUESTED** [review]

File: subsys/bluetooth/mesh/net.c (hunk 3 above)
Comment (Balaklaka) [reply subsys/bluetooth/mesh/net.c:261]: iv_upate is only false here if "missed the Secure Network beacons with the IV update Flag set to 1". Then this have to wait 96 hours before it will be updated, and can come back into the network. That is why the 96 hours rule is left out.

File: subsys/bluetooth/mesh/net.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (omkar3141) [reply subsys/bluetooth/mesh/net.c]: @alxelax  I don't understand how this rule is relevant here. The text you quoted is related to provisioning. In that text, the word "added" means "provisioned".

File: subsys/bluetooth/mesh/net.c (hunk 3 above)
Comment (omkar3141) [reply subsys/bluetooth/mesh/net.c:261]: >this condition requires to be 'and' with 96 hours rule checking 

No, this is precisely what we discussed and we don't want. This situation does not arise if the node is continuously operating in a network. This situation arises: when node is powered off for many hours or days before coming alive back; meanwhile, rest of the network has moved to the next IV index (old iv + 1, and in Normal operation).

File: subsys/bluetooth/mesh/net.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (alxelax) [reply subsys/bluetooth/mesh/net.c]: @omkar3141  how did you determine here this 96 hours limitation is not acceptable here? Is there any attribute or flag that can guarantee 96 hours expired since the last IV Update procedure? Could you point it?

It seems the current implementation will align IV index of the whole network to the device with the highest IV number even device behaves incorrectly. The fact the devices have procedure time limitations will be ignored.

Even more, the specification requires quite strictly time limitation in one more place:
```
3.10.6 IV Index Recovery procedure
The node shall not execute more than one IV Index Recovery within a period of 192 hours.
```
That is not checked anywhere.  Probably it is out of this task since it has not been implemented yet.

**Review by omkar3141 — APPROVED** [review]

**Comment by trond-snekvik** [comment]
Marked with DNM and Enhancement, as this functionality isn't currently broken, and the change needs some more discussion and work. 

The current implementation does not break the spec for the IV update procedure, but it's not very user friendly. The spec is deliberately vague on this point, essentially leaving the IV+1 scenario up to the implementation. 
The options permitted by the spec are:
1. Ignore all non-update IV+1 beacons (current implementation). This ensures that the network will never start galloping through IV indexes because of some broken implementations, but devices that miss out on IV updates because of a power loss will be unable to receive messages from others until the next IV update is triggered by someone else on the network.
2. Treat this as IV recovery, and just adopt the new IV index, skipping update mode and ignoring the 96H timer altogether. This is the implementation initially proposed in this PR, but this opens up the network to start bumping IV index with no timing limitations, allowing broken implementations to exhaust the IV indexes, bricking the network.
3. Adopt the IV index if the 96H limit has been met. This prevents IV index galloping, but we'll still have devices being blocked from participating in the network for up to 96H, like in option 1.
4. Treat this as IV recovery, and just adopt the new IV index, but start enforcing the 192H limit on IV recovery described in section 3.11.6.  This should prevent IV index galloping, and would allow devices to recover immediately in most scenarios. The implementation for this is a bit more complex though, and we have to decide whether the 192H limit be stored persistently, or be reset between power cycles, so devices with frequent power loss could always get back into the network.

After a call between @Balaklaka, @alxelax, @omkar3141 and @trond-snekvik, we figured that option 4 is likely the most usable interpretation, but the implementation details aren't quite clear. Since the current implementation is passing qualification, and not breaking the specification, we decided to not push this into 2.7, and instead revisit after code freeze is lifted, so we can do a proper implementation.

cc @LingaoM and @jhedberg

**Review by omkar3141 — APPROVED** [review]

File: subsys/bluetooth/mesh/net.c [hunk 4]
```diff
@@ -257,24 +267,25 @@ bool bt_mesh_net_iv_update(uint32_t iv_index, bool iv_update)
...
+			if ((ivi_recovery != 0) &&
+			    (k_uptime_delta(&ivi_recovery) < IVI_RECOVERY_TIMEOUT)) {
+				BT_ERR("IV Index Recovery before minimum delay");
+				return false;
+			}
+			/* The spec opens for both of these conditions to
+			 * initiate an IV Index Recovery procedure. It means
+			 * previous IV update has been missed, and node should
+			 * allow the node to continue to function.
+			 */
```
Comment (omkar3141) [inline subsys/bluetooth/mesh/net.c:281]:
```suggestion
			/* The Mesh profile specification allows to initiate an IV 
			 * Index Recovery procedure if previous IV update has
			 * been missed. This allows the node to remain functional.
			 */
```

File: subsys/bluetooth/mesh/net.c [hunk 5]
```diff
@@ -257,24 +267,25 @@ bool bt_mesh_net_iv_update(uint32_t iv_index, bool iv_update)
 			return false;
 		}
 
-		if (iv_index > bt_mesh.iv_index + 1) {
+		if ((iv_index > bt_mesh.iv_index + 1) ||
+		    (iv_index == bt_mesh.iv_index + 1 && !iv_update)) {
+			if ((ivi_recovery != 0) &&
```
[the remark this answers is not in the document]
Comment (alxelax) [reply subsys/bluetooth/mesh/net.c:272]: Discussed IRL with @Balaklaka and @omkar3141 
Seems no big sense to wait until ending pending segmented transaction to start the recovery procedure.
Nature of segmented transaction in BLE Mesh assumes sending all segments and waiting for acknowledgment with confirmation flags. The started transaction doesn't mean the opposite device understood your segments. Better to not wait for timeout expiration to start the recovery procedure.

**Review by alxelax — APPROVED** [review]

**Comment by cfriedt** [comment]
@trond-snekvik - are you able to review?

File: subsys/bluetooth/mesh/net.c [hunk 6]
```diff
@@ -89,6 +96,9 @@ struct bt_mesh_net bt_mesh = {
 	.local_queue = SYS_SLIST_STATIC_INIT(&bt_mesh.local_queue),
 };
 
+/* Time when last IV Index Recovery was done. */
+static int64_t ivi_recovery;
```
Comment (trond-snekvik) [inline subsys/bluetooth/mesh/net.c:100]: Since this value is only set when performing an IV update, and that already runs on a 96+96H time limit, would it be enough to just keep a flag here? e.g. `IV_INDEX_WAS_RECOVERED`, and reset that to false if we change the IV index in a normal way?

**Review by trond-snekvik — CHANGES_REQUESTED** [review]
Should this also be stored persistently? Devices that keep power cycling will not be able to keep track of this limit with this implementation. If this can be made into a flag, it might be pretty cheap to store

**Comment by omkar3141** [comment]
> Should this also be stored persistently? Devices that keep power cycling will not be able to keep track of this limit with this implementation. If this can be made into a flag, it might be pretty cheap to store

This should not be stored persistently that is the central point here. We discussed this in detail in the past. The point is to allow the device which has missed the IV update and got rebooted (i.e. went offline for several hours) to recover IV. The device which reboots quickly will not fall into this condition because its last known IV will be in sync with the network.

File: subsys/bluetooth/mesh/net.c (hunk 6 above)
Comment (Balaklaka) [reply subsys/bluetooth/mesh/net.c:100]: I really like this idea, but is not the IV update timer configurable. So in real it can be longer, but maybe that does not matter.

Comment (trond-snekvik) [reply subsys/bluetooth/mesh/net.c:100]: Yeah, like the 96H IV update delay, we only need the minimum time to pass

Comment (omkar3141) [reply subsys/bluetooth/mesh/net.c:100]: This will again create an issue because the iv_duration timer is intermittently stored in the flash. 

Just to recap what we discussed in the past. The device needs to have the ability to recover the IV index if it is offline when the rest of the network has progressed to the last known IV + 1. There is also a spec requirement that IV recovery can happen only once every 192 hours. When the device is powered off, it loses the track of time in the current stack implementation (although for overall IV procedure there is a timer which is intermittently stored in flash. But under a certain pattern of off-on cycles, even this timer will lag the rest of the network eventually causing the device to fall out of the IV update cycle). With that assumption, the change in this PR allows the device to recover IV exactly once every 192 hours when the device is active without rebooting. This meets the spec constraint for IV recovery. If the device is offline for a brief period of time, the recovery procedure will anyway not trigger because the IV known by the device would match with the IV of the network.

In short, the ivi_recovery should stay volatile and remain separate from iv.iv_duration.

Comment (Balaklaka) [reply subsys/bluetooth/mesh/net.c]: ivi_was_recovered is not stored and will be reset on reboot. So the functionality is still the same.

**Review by trond-snekvik — APPROVED** [review]
